src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,191 @@
+                    var directory = Path.GetDirectoryName(targetPath);
+                    Directory.CreateDirectory(directory);
+                    await AsyncHttpRetryHelper.InvokeVoidAsync(
+                       async () =>
+                       {
+                           using (var sourceStream = await this.DownloadFileFromContainerAsync(containerIdAndRoot, projectId, containerClient, item, cancellationToken))
+                           {
+                               tracer.Info($"Downloading: {targetPath}");
+                               using (var targetStream = new FileStream(targetPath, FileMode.Create))
+                               {
+                                   sourceStream.CopyTo(targetStream);
+                               }
+                           }
+                       },
+                        maxRetries: 5,
+                        cancellationToken: cancellationToken,
+                        tracer: tracer,
+                        continueOnCapturedContext: false,
+                        canRetryDelegate: exception => exception is IOException,
+                        context: null
+                        );
+                },
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = 5000,
+                    MaxDegreeOfParallelism = 8,
+                    CancellationToken = cancellationToken,
+                });
+
+                await downloadBlock.SendAllAndCompleteSingleBlockNetworkAsync(fileItems, cancellationToken);
+        }
+
+        private async Task<Stream> DownloadFileFromContainerAsync(
+            (long, string) containerIdAndRoot,
+            Guid scopeIdentifier,
+            FileContainerHttpClient containerClient,
+            FileContainerItem item,
+            CancellationToken cancellationToken)
+        {
+            Stream responseStream = await AsyncHttpRetryHelper.InvokeAsync(
+                async () =>
+                {
+                    Stream internalResponseStream = await containerClient.DownloadFileAsync(containerIdAndRoot.Item1, item.Path, cancellationToken, scopeIdentifier).ConfigureAwait(false);
+                    return internalResponseStream;
+                },
+                maxRetries: 5,
+                cancellationToken: cancellationToken,
+                tracer: this.tracer,
+                continueOnCapturedContext: false
+                );
+
+            return responseStream;
+        }
+
+        private string ResolveTargetPath(string rootPath, FileContainerItem item, string artifactName)
+        {
+            //Example of item.Path&artifactName: item.Path = "drop3", "drop3/HelloWorld.exe"; artifactName = "drop3"
+            var tempArtifactName = (item.Path.Length > artifactName.Length) ? artifactName + "/" : artifactName;
+            var itemPathWithoutDirectoryPrefix = item.Path.Replace(tempArtifactName, String.Empty);
```

Comment:
I'm confused as to when this would ever apply, the artifact name is part of the item path isn't it?